README.md
# hicat :cat:

![image](https://user-images.githubusercontent.com/74385/102151429-f6ccdd80-3ec6-11eb-84a4-68358ac2bc3b.png)

`cat` with syntax highlighting. The language is auto-detected through the file
extension.

```sh
hicat index.js
```

Pipe something to `hicat`. The language will be inferred from the contents.

```sh
curl http://example.com | hicat
```

If hicat fails to detect a language, specify it using `-l LANG`.

```sh
curl http://example.com | hicat -l xml
```

[![Status](https://travis-ci.org/rstacruz/hicat.svg?branch=master)](https://travis-ci.org/rstacruz/hicat)

## Installation

```sh
npm install -g hicat
```

[![npm version](https://badge.fury.io/js/hicat.svg)](https://npmjs.org/package/hicat "View this project on npm")

Usage:

    $ hicat --help

      Usage:
          hicat [options] FILE
          ... | hicat [options]

      Options:
          -h, --help         print usage information
          -v, --version      show version info and exit
          -l, --lang LANG    use a given language
              --languages    list available languages
              --no-pager     disable the pager

## Tips and tricks

Add an alias to your `~/.bashrc` to save a few keystrokes.

```sh
alias hi=hicat
```

## Btw

[highlight.js] powers the syntax highlighter engine.

## Thanks

**hicat** © 2014+, Rico Sta. Cruz. Released under the [MIT License].<br>
Authored and maintained by Rico Sta. Cruz with help from [contributors].

> [ricostacruz.com](http://ricostacruz.com) &nbsp;&middot;&nbsp;
> GitHub [@rstacruz](https://github.com/rstacruz) &nbsp;&middot;&nbsp;
> Twitter [@rstacruz](https://twitter.com/rstacruz)

[mit license]: https://mit-license.org/
[contributors]: https://github.com/rstacruz/hicat/contributors
[highlight.js]: https://highlightjs.org


History.md
## v0.7.0

 * Automatically prettifies JSON files.

## v0.6.2 - November 10, 2014

 * Add *repository* field to `package.json`. No functional changes.

## v0.6.1 - August 4, 2014

 * Fix `--numbers --debug` adding line numbers to debug info.
 * Fix `--numbers` putting an extra line
 * Always assume that small files (<24 lines) don't need a pager.
 * Don't wrap long lines when using the pager.

## v0.6.0 - August 4, 2014

 * Add the `--numbers` option for line numbers.

## v0.5.1 - August 3, 2014

 * Fix the pager not being on by default.

## v0.5.0 - July 31, 2014

 * Change comment color to cyan.
 * Add colors for javadoc tags.
 * (Internal) documentations and tests.

### v0.4.1

 * JSON: prevent symbols (`{ }`) from being colorized as green.

### v0.4.0

 * Add colors for javadoc comments.
 * Add support for Ruby string interpolation.
 * Make the output of the undocumented `--debug` switch more useful.
 * (Internal) Refactors to tests, color schemes, and many others.

### v0.3.2

 * Add colors for diffs (`--lang diff`).

## v0.3.1 - July 30, 2014

 * Fix piping not working.

### v0.3.0

 * Automatically page the output if the file is long.
 * Throw less-verbose error messages.

### v0.2.3

 * Add colors for Markdown files (`--lang markdown`).
 * Add colors for CSS tokens (`--lang css`).
 * Add colors for C pre-processor directives.

### v0.2.2

 * Fix `--lang` not working properly.

### v0.2.0

 * Add `--languages` option to list available languages.
 * Rename `--type` to `--lang`.

## v0.1.0 - July 30, 2014

 * Initial release.


test/cli.js
require('./setup');

var run = require('./cli_helpers').run,
    pipe = require('./cli_helpers').pipe;
    success = require('./cli_helpers').success;

describe('pipe', function () {

  var input = 'var x = 2';
  pipe(input);
  success();

  it('has no stderr', function () {
    expect(result.stderr).eql('');
  });

  it('has results', function () {
    expect(result.out).eql(hicat(input).ansi);
  });
});

describe('--help', function () {
  run('--help');
  success();

  it('shows --help', function () {
    expect(result.out).include('-h, --help');
    expect(result.out).include('print usage information');
  });
});

describe('--version', function () {
  run('--version');
  success();

  it('shows version info', function () {
    expect(result.out).include(require('../package.json').version);
  });
});

describe('-v', function () {
  run('-v');
  success();

  it('shows version info', function () {
    expect(result.out).include(require('../package.json').version);
  });
});

describe('--languages', function () {
  run('--languages');
  success();

  it('has color constants', function () {
    var out = hicat.listLanguages().sort().join('\n');
    expect(result.out.trim()).eql(out);
  });
});

describe('--list-colors', function () {
  run('--list-colors');
  success();

  it('has color constants', function () {
    expect(result.out).match(/RED=[0-9;]+/);
    expect(result.out).match(/GREEN=[0-9;]+/);
    expect(result.out).match(/BLUE=[0-9;]+/);
    expect(result.out).match(/CYAN=[0-9;]+/);
    expect(result.out).match(/YELLOW=[0-9;]+/);
  });

  it('has "value"', function () {
    expect(result.out).match(/value=.*/);
  });
});

describe('a ruby example', function () {
  run('samples/ruby.rb --no-pager');
  success();

  it('highlights', function () {
    expect(result.out).match(/method/);
    expect(result.out).match(/string/);
  });
});

describe('coercing via --lang', function () {
  run('samples/ruby.rb --no-pager --lang markdown --debug');
  success();

  it('works', function () {
    expect(result.out).match(/language: markdown/);
  });
});

describe('--numbers', function () {
  run('samples/ruby.rb --numbers');
  success();

  it('prefixes with numbers', function () {
    expect(result.stripped).match(/1 class MyClass/);
    expect(result.stripped).match(/2   NUMBER = 200/);
    expect(result.stripped).match(/4   def method/);
  });

  it('colors the numbers', function () {
    expect(result.out).include(hicat.colorize('3 ', hicat.color('line_number')));
  });
});

describe('coercing via -l', function () {
  run('samples/ruby.rb --no-pager -l markdown --debug');
  success();

  it('works', function () {
    expect(result.out).match(/language: markdown/);
  });
});

describe('a ruby example with --debug', function () {
  run('samples/ruby.rb --debug');
  success();

  it('highlights tags', function () {
    expect(result.out).match(/\[string\].*\[\/string\]/);
  });

  it('tells the language', function () {
    var strip = result.out.replace(/\033[^m]*m/g, '');
    expect(strip).include('hicat language: rb');
  });
});

describe('not found', function () {
  run('xxx yyy zzz --no-pager');

  it('fails', function () {
    expect(result.code).eql(8);
  });

  it('does no output', function () {
    expect(result.out).eql('');
  });

  it('reports errors', function () {
    expect(result.stderr).match(/xxx/);
    expect(result.stderr).match(/yyy/);
    expect(result.stderr).match(/zzz/);
    expect(result.stderr).match(/no such file or directory/);
  });
});


test/read.js
/* jshint expr: true */
var expect = require('chai').expect;
var read = require('../lib/read');

describe('read.files', function () {
  var data;

  beforeEach(function (next) {
    read.files(['./test/read.js', './foo'], function (err, _data) {
      expect(err).falsy;
      data = _data;
      next();
    });
  });

  it('returns .files', function () {
    expect(data.files).array;
    expect(data.files).have.length(2);
  });

  it('gives data', function () {
    var file = data.files[0];
    expect(file.data).a('string');
    expect(file.data).match(/Hola mundo/);
    expect(file.file).eql('./test/read.js');
  });

  it('gives errors', function () {
    var file = data.files[1];
    expect(file.error).instanceOf(Error);
    expect(file.error.code).eql('ENOENT');
  });
});

/* Hola mundo */


test/test.js
require('./setup');
var extend = require('util')._extend;

describe('Hicat the module', function() {
  it('is a function', function () {
    expect(hicat).be.a('function');
  });

  it('has colors', function () {
    expect(hicat.colors).be.a('object');
    expect(hicat.colors.string).be.a('string');
  });
});

describe('Explicitly setting the language', function() {
  var out, input;

  beforeEach(function () {
    input = 'var x = 2 + "hi"; /*yo*/';
    out = hicat(input, { lang: 'javascript' });
  });

  it('produces .language', function () {
    expect(out.language).be.eql('javascript');
  });
});

describe('Explicitly setting an invalid language', function() {
  var str, out, input;

  beforeEach(function () {
    input = '{"a":2}';
    out = hicat(input, { lang: 'aoeu' });
  });

  it('auto-detects the language', function () {
    expect(out.language).be.eql('json');
  });
});

describe('A simple example', function() {
  var str, out, input;

  beforeEach(function () {
    input = 'var x = 2345 + "hi"; /*yo*/ window.document.x = function($){};';
    out = hicat(input);
    str = out.ansi;
  });

  it('produces .language', function () {
    expect(out.language).be.eql('javascript');
  });

  it('produces .ansi', function () {
    expect(out.ansi).be.a('string');
  });

  it('produces .raw', function () {
    expect(out.raw).eql(input);
  });

  it('produces ansi codes', function () {
    var code = str.replace(/\033[^m]+m/g, '');
    expect(code).eql(input);
  });

  it('highlights numbers', function () {
    var fragment = hicat.colorize('2345', hicat.color('number', 'javascript'));
    expect(str).include(str);
  });

  it('highlights comments', function () {
    var fragment = hicat.colorize('/*yo*/', hicat.color('comment', 'javascript'));
    expect(str).include(fragment);
  });
});

describe('Hicat.color', function () {
  var oldcolors;

  beforeEach(function () {
    oldcolors = extend({}, hicat.colors);
  });

  afterEach(function () {
    hicat.colors = oldcolors;
  });

  it('works for simple cases', function () {
    hicat.colors = { string: '80' };
    expect(hicat.color('string')).eql('80');
  });

  it('resolves references', function () {
    hicat.colors = { string: '70', str: 'string', s: 'str' };
    expect(hicat.color('s')).eql('70');
  });

  it('accounts for languages', function () {
    hicat.colors = { val: '20', 'json:val': '30' };
    expect(hicat.color('val')).eql('20');
    expect(hicat.color('json:val')).eql('30');
  });

});

describe('Hicat.colorize', function () {
  it('works', function () {
    var fragment = hicat.colorize('yo', '32');
    expect(fragment).eql('\033[32m' + 'yo' + '\033[0m');
  });
});


index.js
var hljs = require('highlight.js');
var colorize;

/***
 * hicat() : hicat(str, options)
 * Highlights a given `str` string.
 *
 *   var hicat = require('hicat');
 *   hicat("echo 'hi'", { filename: 'script.sh' })
 *   => "echo \033[32m'hi'\033[0m"
 *
 * Options available:
 *
 * ~ filename (String): filename
 * ~ lang (String): language to use
 * ~ debug (Boolean): shows extra info if `true`
 * ~ numbers (Boolean): shows line numbers if `true`
 */

function hicat (str, options) {
  if (!options) options = {};

  var lang = options.lang || (options.filename && extname(options.filename));
  var out;

  if (lang) {
    try {
      out = hljs.highlight(lang, str);
    } catch (e) {
      out = hljs.highlightAuto(str);
    }
  } else {
    out = hljs.highlightAuto(str);
  }

  if (!out || !out.value) throw new Error("failed to highlight");

  // convert <span> tags to ANSI colors
  out.ansi = html2ansi(out.value, {
    lang: out.language,
    debug: options.debug
  });

  // Add line numbers
  if (options.numbers) {
    var i = 0;
    var lines = out.value.split("\n").length;
    var digits = (""+lines).length;
    if (digits < 3) digits = 3;
    out.ansi = out.ansi.replace(/^/gm, function (s) {
      i++;
      if (i >= lines) return s;

      var pad = digits - (""+i).length;
      var prefix = Array(pad+1).join(" ") + i + ' ';
      prefix = colorize(prefix, color('line_number'));
      return prefix + s;
    });
  }

  // Add language comment
  if (options.debug) {
    var info = "/* hicat language: " + out.language + " */";
    info = colorize(info, color('tag', 'debug'));
    out.ansi += "\n\n" + info + "\n";
  }

  return {
    ansi: out.ansi,
    language: out.language,
    html: out.value,
    raw: str
  };
}

/**
 * colors : hicat.colors
 * The color scheme. This is a key-value object that `hicat.color()` refers to.
 */

hicat.colors = require('./lib/colors');

/**
 * listLanguages() : hicat.listLanguages()
 * Returns a list of supported languages.
 *
 *   listLanguages()
 *   => ['javascript', 'python', 'c', ...]
 */

hicat.listLanguages = hljs.listLanguages;

/**
 * extname() : extname(filename)
 * (private) Extracts the extension from a given `filename`.
 *
 *   extname('hi.json')
 *   => 'json'
 */

function extname (fname) {
  var m = fname.match(/\.([^\.]+)$/);
  if (m) return m[1];
}

/**
 * color() : hicat.color(token)
 * Returns the color for a given token.
 *
 *     color('string')
 *     => '32'
 *     color('attribute', 'json')
 *     => '32'
 */

var color = hicat.color = function (token, lang) {
  if (lang)
    return getColor(lang + ':' + token) || getColor(token);
  else
    return getColor(token);

  function getColor (token) {
    var code = token, newcode;
    while (true) {
      newcode = hicat.colors[code];
      if (newcode) code = newcode;
      else if (token !== code) return code;
      else return;
    }
  }
};

/**
 * html2ansi() : html2ansi(str, options)
 * (private) Converts hljs-style spans from a given HTML `str` into ANSI
 * color codes. Available options:
 *
 * ~ lang (String): language
 * ~ debug (Boolean): true or false
 *
 *   html2ansi('<span class="hljs-string">"hi"</span>")
 *   => "\033[31m"hi"\033[0m"
 */

function html2ansi (str, options) {
  // do multiple passes as spans can be multiple
  while (~str.indexOf('<span class="hljs-')) {
    str = replaceSpan(str, options);
  }

  return str
    .replace(/<span class="([^"]*)">/g, '')
    .replace(/<\/span>/g, '')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&');
}

/**
 * replaceSpan() : replaceSpan(html, options)
 * (private) Replaces span tags with ANSI codes in the given `html` string.
 * A delegate of `html2ansi`.
 *
 *   html = '<span class="hljs-tag">hi</span>'
 *   replaceSpan(html, 'java')
 *   => "\033[0;32mhi\033[0m"
 */

function replaceSpan (str, options) {
  return str
    .replace(/<span class="hljs-([^"]*)">([^<]*)<\/span>/g, function (_, token, s) {
      var code = color(token, options && options.lang);
      if (options.debug) {
        return "" +
          colorize("[" + token + "]", color('tag', 'debug')) +
          colorize(s, code) +
          colorize("[/" + token + "]", color('tag', 'debug'));
      }
      return colorize(s, code);
    });
}

/**
 * colorize() : hicat.colorize(str, color)
 * Applies the color `color` to the string `str`.
 *
 *   colorize("hello", 32)
 *   => "\033[32mhello\033[0m"
 */

colorize = hicat.colorize = function (s, color) {
  if (!color) return s;

  var reset = "\033[0m",
      code = "\033[" + color + "m";

  // improved support for less -R, since it automatically resets the colors per line
  if (~s.indexOf("\n")) s = s.replace(/\n/g, "\n"+code);

  // nesting
  if (~s.indexOf(reset)) {
    s = s.replace(/\033\[0m/g, code);
  }

  return code + s + reset;
};

module.exports = hicat;


lib/read.js
var fs = require('fs');

/**
 * read() : read(files, fn)
 * Reads from files. If no files are given, read from stdin.
 * The `err` argument will always be null, as errors will be part of `res`.
 *
 *     fnames = process.argv.slice(2); //=> ['readme.txt']
 *
 *     read(fnames, function (err, res) {
 *       res.data    //=> '...'
 *       res.error   //=> undefined | Error()
 *       res.stdin   //=> true | false
 *       res.files   //=> [...]
 *
 *       res.files.forEach(function (f) {
 *         f.data    //=> ...
 *         f.error   //=> undefined | Error(...)
 *         f.stdin   //=> true | false
 *         f.name    //=> 'readme.txt'
 *       }
 *     });
 *
 * If `files` is a blank array (or null), data will be read from stdin. The
 * resulting data will have a similar schema.
 *
 *     read([], function (err, res) {
 *       ...
 *     });
 */

function read (files, fn) {
  // from stdin
  if (!files || files.length === 0) {
    read.stdin(function (err, data) {
      if (err)
        fn(null, new Result([{ stdin: true, error: err }]));
      else
        fn(null, new Result([{ stdin: true, data: data }]));
    });
  }
  // from files
  else {
    var out = files.map(function (fname) {
      try {
        var data = fs.readFileSync(fname, 'utf-8');
        return { name: fname, data: data };
      } catch (err) {
        return { name: fname, error: err };
      }
    });

    out = new Result(out);
    fn(null, out);
  }
}

/**
 * read.stdin() : read.stdin(fn)
 * Read data from standard input. The `err` argument will always be null.
 *
 *   read.stdin(function (err, data) {
 *     console.log(data); // string
 *   });
 */

read.stdin = function (fn) {
  var data = '';

  process.stdin.setEncoding('utf8');

  process.stdin.on('readable', function() {
    var chunk = process.stdin.read();
    if (chunk !== null) data += chunk;
  });

  process.stdin.on('end', function() {
    fn(null, data);
  });
};

/**
 * res:
 * The results value is an object passed to the callback of `read()`.
 *
 * ~ data (String): a concatenation of all data in all the files.
 * ~ error (Error): The first error in all files. `undefined` if successful.
 * ~ stdin (Boolean): is `true` if the file is read from stdin
 * ~ files (Array): A list of files.
 *
 * Each of the items in `files` has a similar list of values:
 *
 * ~ data (String): File data
 * ~ error (Error): error, if applicable
 * ~ stdin (Boolean): is `true` if the file is read from stdin
 * ~ name (String): File name
 *
 * See [read()](read) for an example.
 */

function Result(files) {
  this.files = files;
}

getter(Result.prototype, 'data', function () {
  return this.files.map(function (f) { return f.data || ""; }).join("");
});


getter(Result.prototype, 'error', function () {
  return this.files.reduce(function (acc, f) { return acc || f.error; });
});

getter(Result.prototype, 'stdin', function () {
  return this.files && this.files[0] && this.files[0].stdin;
});

/**
 * getter() : getter(prototype, prop, fn)
 * (private) Defines a get property `prop` in the given `prototype` object.
 */

function getter (proto, prop, fn) {
  Object.defineProperty(proto, prop, { get: fn });
}

module.exports = read;


lib/term.js
/**
 * print() : print(str)
 * Prints the given string `str` to the terminal. If it is too long, it is ran
 * through a pager.
 */

exports.print = function (str) {
  var count = str.split("\n").length;
  exports.lines(function (err, max) {
    if (max && count > max) {
      exports.page(str);
    } else {
      process.stdout.write(str);
    }
  });
};

/**
 * lines() : lines(fn)
 * (private) Returns the number of lines in the terminal.
 *
 *   lines(function (err, n) {
 *     n == 24
 *   });
 */

exports.lines = function (fn) {
  require('child_process').exec('tput lines', function (err, cout, cerr) {
    if (err) fn();
    else fn(null, +cout);
  });
};

/**
 * page() : page(str)
 * (private) Prints the `str` through a pager.
 */

exports.page = function (str) {
  var spawn = require('child_process').spawn;
  var child = spawn('less', ['-R'], { stdio: [ 'pipe', 1, 2 ] });
  child.stdin.write(str);
  child.stdin.end();
};


lib/colors.js
/*
 * Color scheme
 */

module.exports = {
  'GREY': '0;30',
  'RED': '0;31',
  'GREEN': '0;32',
  'YELLOW': '0;33',
  'BLUE': '0;34',
  'MAGENTA': '0;35',
  'CYAN': '0;36',
  'BOLD': '0;1',
  'UNDERLINE': '0;4',
  'BOLDWHITE': '1;37',
  'NIL': '0;0',

  'keyword': 'BOLD',            // like 'class', 'module', etc
  'built_in': 'keyword',        // like var, etc
  'tag': 'keyword',             // css tag
  'class': 'keyword',           // css class

  'title': 'UNDERLINE',         // tags, function names

  'comment': 'CYAN',
  'doctype': 'comment',
  'pi': 'comment',              // xml declaration
  'preprocessor': 'comment',    // C/C++ preprocessor directives
  'horizontal_rule': 'comment', // markdown HR ---
  'javadoc': 'comment',         // double-star comments
  'blockquote': 'comment',      // markdown blockquote
  'javadoctag': 'BLUE',         // tags like @param

  'string': 'GREEN',
  'value': 'string',            // html/json values
  'json:value': 'NIL',
  'code': 'string',             // markdown code
  'link_reference': 'string',   // markdown link reference

  'number': 'YELLOW',
  'symbol': 'number',           // ruby ':symbols
  'hexcolor': 'number',         // css hex holor
  'link_url': 'number',         // markdown link url

  'function': 'BOLD',           // css function, like rgba()
  'strong': 'function',         // markdown strong
  'link_label': 'strong',       // markdown links

  'header': 'BOLDWHITE',        // markdown header

  'attribute': 'BLUE',          // html/json attributes
  'literal': 'attribute',       // true, false, etc
  'variable': 'attribute',      // xml variable

  'pseudo': 'GREY',             // css pseudoclass :before
  'attr_selector': 'pseudo',    // css attr [type='number']

  'params': 'NIL',              // function parameters ()
  'regexp': 'MAGENTA',

  'subst': 'NIL',               // ruby interpolation

  'addition': 'GREEN',          // diff +
  'deletion': 'RED',            // diff -
  'chunk': 'BLUE',              // diff @@ -1,119 +1,768 @@

  'debug:tag': 'GREY'
};
